packages/notebook/src/aviary/envs/notebook/utils.py

```diff
@@ -0,0 +1,491 @@
+                    stats["comment_lines"] += 1
+                else:
+                    stats["code_lines"] += 1
+
+            # Check for R and bash cells
+            if lines:
+                first_line = lines[0].strip()
+                if first_line.startswith("%%R"):
+                    stats["r_cells"] += 1
+                elif first_line.startswith("%%bash"):
+                    stats["bash_cells"] += 1
+
+                # Count shell commands (lines starting with !)
+                stats["shell_commands"] += sum(
+                    1 for line in lines if line.strip().startswith("!")
+                )
+
+            # Check outputs for images and tables
+            if hasattr(cell, "outputs"):
+                for output in cell.outputs:
+                    # Check for images
+                    if output.get("output_type") in {"display_data", "execute_result"}:
+                        if "image/png" in output.get("data", {}):
+                            stats["images"] += 1
+
+                        # Check for HTML tables or DataFrame representations
+                        if "text/html" in output.get("data", {}):
+                            html_content = output["data"]["text/html"]
+                            if isinstance(html_content, list):
+                                html_content = "".join(html_content)
+                            if "<table" in html_content:
+                                stats["tables"] += 1
+
+                        # Check for plain text DataFrame representations
+                        elif "text/plain" in output.get("data", {}):
+                            text_content = output["data"]["text/plain"]
+                            if isinstance(text_content, list):
+                                text_content = "".join(text_content)
+                            if any(
+                                marker in text_content
+                                for marker in ("DataFrame", "Series")
+                            ):
+                                stats["tables"] += 1
+
+        elif cell.cell_type == "markdown":
+            stats["markdown_lines"] += len(lines)
+            stats["markdown_cells"] += 1
+
+            # Count markdown images
+            for line in lines:
+                if "![" in line or "<img" in line:
+                    stats["images"] += 1
+    return stats
+
+
+def extract_xml_content(text: str, tag_name: str) -> str | None:
+    """
+    Extract content between XML-like tags from text.
+
+    Args:
+        text (str): The text containing XML-like tags
+        tag_name (str): The name of the tag to extract content from
+
+    Returns:
+        str or None: The content between the tags, or None if not found
+    """
+    pattern = f"<{tag_name}>(.*?)</{tag_name}>"
+    match = re.search(pattern, text, re.DOTALL)
+
+    if match:
+        return match.group(1).strip()
+    return None
+
+
+def litellm_wrapper(
```

Comment:
If we used `lmi` over this, we could drop `litellm` as a primary dep for this package

And feel free to ignore this one